### ISSUE TITLE
Support Logger for debug log messages

### ISSUE THREAD

**Issue by dannypurcell** [LWJGL/lwjgl3#67]
I love the support for debugging in GLContext. The setupDebugMessageCallback() method is super convenient.
While building up the debugging support in my own project I thought of something that would make the logging support even better.
Please consider adding a method that sends debug messages to a java.util.logging.Logger rather than a PrintStream. This would help lwjgl users easily tie GL debug messages in to application logs and enable the output during runtime.

To take this idea a bit further, the holy grail would be swapping in [slf4j](http://www.slf4j.org/) loggers for log output throughout the lwjgl library. That might take a bit more time to implement though so feel free to consider slf4j a low priority nice to have. 
java.util Loggers are already on hand and slf4j can remap their output so that might be the best option if adding the slf4j dependency is undesirable.

Thanks again for all the great work! The new version is looking really good!

**Comment by Spasi** [comment]
I will wait for input from other users before doing anything. But I don't think having a dependency on slf4j would be a good idea (for LWJGL users, I personally don't mind).

**Comment by httpdigest** [comment]
Personally, I would like to have more control over the formatting of the generated messages.
And this would usually be done with a common callback interface that the user can set.
Having had a look into the relevant code in LWJGL 3 the problem seems, though, that the callback interface and the actual formatting of the messages are intertwined because both depend on the underlying native extension/GL core function being used, which have differing signatures and provided information.
The only thing I can think of right now is a common interface which is the union of the information provided by all different underlying callbacks.
This way the user can adapt any logging library he/she wants to use, without introducing a dependency into LWJGL 3.

**Comment by Spasi** [comment]
Well, `setupDebugMessageCallback` is a "reference implementation" really. It is almost too complex for LWJGL, which is supposed to be super lightweight and not provide high-level abstractions. The only reason it's there is because it's so damn useful. My experience so far has been that many users simply ignore that the modern debug mechanisms exist and go through the pain of good old `glGetError`.

**Comment by httpdigest** [comment]
> It is almost too complex for LWJGL ...

Yes, it almost feels like a "utility" method. But there is a pattern to observe: The more something becomes "utility" the more people demand for it and the more useful it becomes.
See for example the math classes. :)

> ... and go through the pain of good old glGetError

That's a pitty. But glGetError of course has its use. I once used a very old Intel card with a GL 2.0-only driver, that had neither of these debugging extensions. But of course, I wouldn't develop and debug on that, but glGetError was still necessary to catch errors only occurring there.

**Comment by Spasi** [comment]
> Yes, it almost feels like a "utility" method. But there is a pattern to observe: The more something becomes "utility" the more people demand for it and the more useful it becomes.
See for example the math classes. :)

Of course, but there's nothing bad about making standalone libraries out of specialized utility functionality. JOML is a perfect example of that.

By adding this method to LWJGL, trying to be helpful, I have created more work for myself. Someone asks for Logger/slf4j integration. Then someone else asks for a high-level callback. Who knows what comes next... You cannot keep everyone happy and the only way to satisfy a request is by making the library opinionated, which LWJGL doesn't want to be.

The raw functionality is there, it's simple and publicly available, anyone could roll their own implementation. This is even the primary reason LWJGL 3 exists and the whole point of writing it from scratch.

**Comment by httpdigest** [comment]
I agree with you on all points there. I was actually never in favor of having "utility" things in LWJGL itself, just mentioned that those things are useful. :)
LWJGL is fine as it is and that particular utility method shouldn't be there in the first place.
And yes, clients should roll their own utilities on top of what LWJGL provides as low-level library.
It's just that most clients probably woudn't know how to do that properly, or even that the GL debugging extensions exist. That's probably why that utility method is in GLContext.

**Comment by maatvirtue** [comment]
Personally I would like to be able to customize log message format by having LWJGL using SLF4J.
I am using SLF4J in my project so that would make it very easy to me.

**Comment by Spasi** [comment]
I'm currently experimenting with how to address this issue:

- Any time LWJGL prints debug information or non-fatal errors, it uses a PrintStream instance in LWJGLUtil.
- That instance defaults to System.err.
- It can be overridden with a system property (that specifies a factory class that returns a PrintStream instance).

Using the above, I'm able to create a PrintStream implementation that forwards everything to a logger or whatever. Any comments?

**Comment by dannypurcell** [comment]
I think that sounds workable. My only reservation is that it may be duplicating functionality already provided by java.util.logging.Logger. 
If `LWJGLUtil.log(CharSequence)` could switch to a Logger based on a system property that might cover it without duplication.

**Comment by Spasi** [comment]
> My only reservation is that it may be duplicating functionality already provided by java.util.logging.Logger. 
If LWJGLUtil.log(CharSequence) could switch to a Logger based on a system property that might cover it without duplication.

I wanted to avoid such a solution because SLF4J users would then be forced to use the jul-to-slf4j bridge.

LWJGL now supports (in build 3.0.0b build 37) programmatic configuration with the [org.lwjgl.system.Configuration](https://github.com/LWJGL/lwjgl3/blob/master/modules/core/src/main/java/org/lwjgl/system/Configuration.java) class. It can be used to setup forwarding to any logger with:

```
Configuration.DEBUG.set(true);
Configuration.setDebugStreamConsumer(new DebugStreamConsumer() {
	@Override
	public void accept(String message) {
		Logger.getLogger("MyLogger").info(message);
	}
});
```

I believe that's easy enough and avoids any dependencies to j.u.l.Logger or any other logging framework.

**Comment by mufumbo** [comment]
@Spasi this haven't been released to the maven release 3.1? or changed since there?

**Comment by Spasi** [comment]
Yes, `setDebugStreamConsumer` has been removed. Its implementation was complicated and its usefulness questionable.

You can still use `Configuration.DEBUG_STREAM` to forward messages to a custom logging solution.
